geotrek/cirkwi/parsers.py

```diff
@@ -0,0 +1,302 @@
+    }
+
+    def __init__(self, *args, **kwargs):
+        # Extract URL parameter to use to retrieve updates only (will be used in 'next_row' method)
+        self.updated_after = None
+        if self.update_only and self.model.objects.filter(provider__exact=self.provider).exists():
+            last_update_timestamp = self.model.objects.filter(provider__exact=self.provider).latest(
+                'date_update').date_update.timestamp()
+            self.updated_after = str(int(last_update_timestamp))
+        super().__init__(*args, **kwargs)
+
+    def normalize_field_name(self, name):
+        return name
+
+    def next_row(self):
+        # Get data from local file :
+        if self.filename:
+            with open(self.filename) as f:
+                xml_root = ET.fromstring(f.read())
+            # Yield objects given XML path in 'results_path'
+            entries = xml_root.findall(self.results_path)
+            self.nb = len(entries)
+            for row in entries:
+                yield row
+
+        # Get data from API URL :
+        else:
+            # Make first query to retrieve objects count
+            # We don't need the objects yet, just need to access 'nb_objects', so set params to 0
+            params = {
+                'first': 0,
+                'rows': 0,
+            }
+            if self.updated_after:
+                params['end-time'] = self.updated_after
+            response = self.request_or_retry(self.url, params=params, auth=self.auth)
+            # Save objects count
+            self.nb = int(ET.fromstring(response.content).find("listing_ids", {}).attrib['nb_objects'])
+
+            # Make several requests, using Cirkwi pagination parameters, until all objects are downloaded
+            first = 0
+            while first <= self.nb:
+                params['first'] = first
+                params['rows'] = self.rows
+                response = self.request_or_retry(self.url, params=params, auth=self.auth)
+                xml_root = ET.fromstring(response.content)
+                # Yield objects given XML path in 'results_path'
+                entries = xml_root.findall(self.results_path)
+                for row in entries:
+                    yield row
+                first += self.rows
+
+    def get_part(self, dst, src, val):
+        # Recursively extract XML attributes
+        if '@@' in src and src[:2] != '@@':
+            part, attrib = src.split('@@', 1)
+            return self.get_part(dst, f"@@{attrib}", val.find(part))
+        # Extract XML attributes
+        elif src.startswith('@@'):
+            return val.attrib[src[2:]]
+        else:
+            # Replace language attribute
+            if "'<LANG>'" in src:
+                src = src.replace("<LANG>", self.default_language)
+            # Return a list of XML elements
+            if src.endswith('/*'):
+                return val.findall(src[:-2])
```

Comment:
I think it is possible to return directly strings here instead of a list of XML elements. This way it would work in the same way than when a single text value is returned at line 114. And you could remove all conversion done in `filter_*` methods.

Reply:
It gets tricky with lists here, because, what do we do after retrieving the multiple elements ? Do we need to get the list of inner texts, or a list from the attributes ? Are we looking for one or several other inner element(s) ? We would need to add some more recursivity to combine this with the other syntaxes. It's why I chose to use `get_part` to return the list of elements and let `filter_xxxx` handle further processing for now

Reply:
Alright, let's leave it as is then.